server/controller/src/main/java/ai/starwhale/mlops/datastore/DataStore.java

```diff
@@ -51,69 +54,178 @@ public void update(String tableName,
             }
+            var records = results.stream()
+                    .map(r -> this.encodeRecord(columnTypeMap, r.getValues(), req.isRawResult()))
+                    .collect(Collectors.toList());
+            return new RecordList(columnTypeMap, records, lastKey);
+        } finally {
+            table.unlock();
         }
-        if (iters.isEmpty()) {
-            return new RecordList(null, null, null);
+    }
+
+    public RecordList scan(DataStoreScanRequest req) {
+        var limit = req.getLimit();
+        if (limit > 1000) {
+            throw new SWValidationException(SWValidationException.ValidSubject.DATASTORE,
+                    "limit must be less or equal to 1000. request=" + req);
         }
-        var columnTypeMap = new HashMap<String, ColumnType>();
-        for (var it : iters) {
-            columnTypeMap.putAll(it.getColumnTypeMapping());
+        if (limit < 0) {
+            limit = 1000;
         }
-        var keyColumnType = iters.get(0).getKeyColumnType();
-        Object lastKey = null;
-        List<Map<String, String>> ret = new ArrayList<>();
-        while (!iters.isEmpty() && (req.getLimit() < 0 || ret.size() < req.getLimit())) {
-            lastKey = Collections.min(iters, (a, b) -> {
-                @SuppressWarnings("rawtypes") var x = (Comparable) a.getKey();
-                @SuppressWarnings("rawtypes") var y = (Comparable) b.getKey();
-                //noinspection unchecked
-                return x.compareTo(y);
-            }).getKey();
-            var record = new HashMap<String, String>();
-            for (var iter : iters) {
-                if (iter.getKey().equals(lastKey)) {
-                    record.putAll(iter.getRecord());
-                    iter.next();
+
+        var tablesToLock =
+                req.getTables()
+                        .stream()
+                        .map(DataStoreScanRequest.TableInfo::getTableName)
+                        .sorted() // prevent deadlock
+                        .map(this::getTable)
+                        .collect(Collectors.toList());
+
+        for (var table : tablesToLock) {
+            table.lock();
+        }
+        try {
+            class TableMeta {
+                String tableName;
+                MemoryTable table;
+                TableSchema schema;
+                Map<String, String> columns;
+                Map<String, ColumnType> columnTypeMap;
+            }
+
+            var tables = req.getTables().stream().map(info -> {
+                var ret = new TableMeta();
+                ret.tableName = info.getTableName();
+                ret.table = this.getTable(info.getTableName());
+                ret.schema = ret.table.getSchema();
+                ret.columns = this.getColumnAliases(ret.schema, info.getColumns());
+                ret.columnTypeMap = ret.schema.getColumnTypeMapping(ret.columns);
```

Comment:
It seems that `keepNone` of tableInfo is useless?

Reply:
Done